File: 2019/d02/main.js
const fs = require('fs');
let file = fs.readFileSync('input.txt', {encoding: 'ascii'});

class Intcode {
	constructor(str, noun = null, verb = null) {
		this.program = str;
		this.setup(noun, verb);
		this.instructions = {
			1: (offset) => { this.pointer(offset + 3, this.pointer(offset + 1) + this.pointer(offset + 2)); return 4; },
			2: (offset) => { this.pointer(offset + 3, this.pointer(offset + 1) * this.pointer(offset + 2)); return 4; },
			99: (offset) => { return null; }
		}
	}

	pointer(offset, value = null) {
		if (value === null) {
			return this.code[this.code[offset]];
		} else {
			this.code[this.code[offset]] = value;
			return;
		}
	}

	setup(noun, verb) {
		this.code = [];
		this.position = 0;
		let re = /(\d+)/g;
		let m = null;
		while (m = re.exec(this.program)) {
			this.code.push(parseInt(m[1]));
		}
		if (noun !== null) this.code[1] = noun;
		if (verb !== null) this.code[2] = verb;
	}

	step() {
		let opcode = this.code[this.position];
		let offset = this.instructions[opcode](this.position);
		this.position += offset;
		return offset !== null;
	}

	first() {
		return this.code[0];
	}

	run() {
		while (this.step());
		return this.first();
	}
};

let pc = new Intcode(file, 12, 2);
// let pc = new Intcode("1,1,1,4,99,5,6,0,99");
console.log(pc.run());

File: 2019/d02/main.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <regex>
#include <vector>

class Intcode {
	std::vector<int> m_code;
	std::vector<int>::iterator m_position;
	std::string m_init;

public:
	Intcode(std::string str) {
		m_init = str;
		setup(12, 2);
	}

	void setup(int noun, int verb) {
		m_code.clear();
		std::regex re("(\\d+),?");
		for (std::sregex_iterator it(m_init.begin(), m_init.end(), re), last{}; it != last; it++) {
			m_code.push_back(std::stoi((*it)[1]));
		}
		m_code[1] = noun;
		m_code[2] = verb;
		m_position = m_code.begin();
	}

	bool step() {
		int opcode = *m_position;
		if (opcode == 99) return false;

		int offset1 = *(m_position + 1);
		int offset2 = *(m_position + 2);
		int offset3 = *(m_position + 3);

		switch (opcode) {
			case 1: {
				m_code[offset3] = m_code[offset1] + m_code[offset2];
			}
			break;
			case 2: {
				m_code[offset3] = m_code[offset1] * m_code[offset2];
			}
			break;
		}

		m_position += 4;
	}

	int first() {
		return m_code[0];
	}

	int run() {
		while (step());
		return first();
	}
};

int main() {
	std::ifstream file("input.txt");
	std::stringstream ss;
	ss << file.rdbuf();
	auto str = ss.str();
	Intcode ic(str);

	std::cout << "part 1: " << ic.run() << std::endl;

	int noun = 0, verb = 0;
	do {
		ic.setup(noun++,verb);
		if (noun > 99) { verb++; noun = 0; }
	} while (ic.run() != 19690720);
	noun--;
	std::cout << "part 2: " << 100 * noun + verb << std::endl;
}